fitbenchmarking/fitbenchmarking
# Add summary page for problems

#### Description of Work

Adds a summary page so that cost functions can be more easily compared.

Fixes #491

Summary of main changes:
- **Allow multiple cost functions**
  This is the main reason for the pr. Basically allow users to run fitbenchmarking with multiple cost functions and fix all associated problems...
- **Flatten the list of results returned from core/fitting_benchmarking**
  This is so that the results can be sorted as the display sees fit and further decouples the 2 parts of the software. One possible use for this is that it will be easier to allow users to select what the columns and rows in the table represent.
- **All results functions use a preprocessed dataset**
  This formats the data into a a dictionary of the form `{<row_name>: {<category_name>: [<column_result>, ...], ...}, ...}`. The terminology might be a bit confusing in places so feel free to make me fix them if it is, but the idea is that categories may be something other than cost_functions in the future. Categories introduce a new element to the tables where a category is displayed as new columns but comparisons are restricted to within the category.
- **Summary Page**
  This adds plots to compare category results and can be accessed by clicking the name of the problem on an index page.
- **Remove cost function descriptions**
  The existing method for this was not easily extended and seemed like it would be too much if there were multiple. I suggest these are added back in in a new PR (via the cost_function class in results_processing without passing it back from fitting_benchmarking) with some extra thought on the best way of showing 2 or 3 in the UI.
- *Fixed `minimzers` typo in fitting_benchmarking...*
  This was annoying me when I was editing the lines with it on...


I'm aware that this PR is already huge (apologies to anyone reviewing...) so will likely create new issues for the following when this is merged:
- Add links to support pages for best result from summary page (#?)
- Parse descriptions from problems to show in summary pages (#948)
- Reintroduce cost function descriptions as accordian/tabbed/new page(?) on index page. (#?)
- Add support for log plots in the problem summary page (#?)

#### Testing Instructions

1. Run fitbenchmarking with a single cost function and check the results including the new summary page
2. Run again with multiple cost functions

Function: Does the change do what it's supposed to?
Tests: Does it pass? Is there adequate coverage for new code?
Style: Is the coding style consistent? Is anything overly confusing?
Documentation: Is there a suitable change to documentation for this change?
